### ISSUE TITLE
Add an option to specify the number of messages to retrieve

### ISSUE THREAD

**Issue by fplanjer** [Scooletz/QueueBatch#19]
I see that the number of messages that are retrieved in one call is set to 32 (or CloudQueueMessage.MaxNumberOfMessagesToPeek, which I assume is 32 also).

Do you see any issue in extending the functionality to make this number configurable?

**Comment by Scooletz** [comment]
You are correct, currently the maximum number of messages is retrieved as one batch. Let me provide some reasoning against it:

If you can amortize the cost of accessing another service (for instance a blob storage) by gathering as many messages as possible, you should do it to "smart batch" them which would result in a better throughput and lowering costs of running your solution. If for any reason, you know that there's a limit (for instance some kind of a size limit) then, by iterating over the batch in `foreach` you can easily calculate can you move one forward and acknowledge only these that have been processed so far. There's nothing wrong in issuing another call to the external service with the rest of messages. Does it show why I don't see a need of adjusting it? Maybe you could you provide some reasoning for this change @fplanjer ?

**Comment by fplanjer** [comment]
Thanks for your reply and taking the time to explain. Your reasoning makes sense. 

The reason we are looking into this is that the actual processing of the message is done by an executable which handles at most 5 messages at a time (it is processing images). The reason for 5 images is unknown to me, but this was identified by the team creating this executable to be an efficient number of items, in terms of cpu, memory, etc.

We could indeed just only process the first 5 messages of the batch and not acknowledge the rest, but this seems to be less efficient than only taking the correct number of message to start with. With the numbers we are looking at we need to be as efficient as possible,

**Comment by Scooletz** [comment]
Ok, now I see it. I'm not sure why would you process only first 5 and discard the rest though. Can't you process 5 from the batch and then take another 5 from the batch? Something that would look like

```c#
public static void MyFunc([QueueBatchTrigger("myqueue")] IMessageBatch batch)
{
  foreach (var fiveMessages in batch.Messages.GroupInFives())
  {
    if (TryProcessImages(fiveMessages))
    {
       foreach (var msg in fiveMessages)
       {
          batch.MarkAsProcessed (msg);
       }
    }
  }
}
```

**Comment by tmenier** [comment]
Thank you for this library. I'm evaluating whether it could be useful for my purposes, and upon reading the readme I think the most obvious question was, where do I set the batch size? If it's fixed at 32 then I think that should at least be mentioned, but I agree with @fplanjer that there are plenty of reasons you'd want to configure it. It just seems reasonable to expect a "batching" library to provide this. In your example that takes 32 and breaks into sub-batches of 5, sure, that works, but imagine how much nicer that code would be if you could just add `BatchSize = 5` to the attribute. ;)

In my case, I'm sending data to an external API that has a batch limit of 100, so I'd actually want to increase it. What would be the ultimate killer feature (and I can open this in a new issue if you think you'd consider it) would be the ability to specify both a batch size _and_ a timeout value. This would allow me to specify: "give me up to 100 items, but if a full minute has passed just give me what you have".

Thanks again!

**Comment by Scooletz** [comment]
The `Listener` not only tries to get 32 times at one time (to reduce the overhead for obtaining a limited number) but also does it in parallel

https://github.com/Scooletz/QueueBatch/blob/b2ced5d726c5dd6fa98db52007810b3804772d74/src/QueueBatch/Impl/Listener.cs#L81

This is done to address high throughput scenario. It'd be possible to set `parallelGets` to `1` and then amend the batch size. If you find a way to address it and want to issue a PR @tmenier I can review it in a week or two.
